components/server/src/ome/services/scheduler/ThreadPool.java

```diff
@@ -7,46 +7,173 @@
 
+import java.util.concurrent.Callable;
 import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
 import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.RunnableFuture;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.SynchronousQueue;
 import java.util.concurrent.ThreadFactory;
 import java.util.concurrent.ThreadPoolExecutor;
 import java.util.concurrent.TimeUnit;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import org.quartz.Scheduler;
+import org.perf4j.StopWatch;
+import org.perf4j.slf4j.Slf4JStopWatch;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
 
 /**
- * Produces a <a href="http://www.opensymphony.com/quartz/Quartz</a>
- * {@link Scheduler} which automatically loads all the triggers it can find.
- *
- * @author Josh Moore, josh at glencoesoftware.com
- * @since Beta4.2.1
+ * An internal implementation of {@link ThreadPoolExecutor} that additionally
+ * handles the submission of {@link #background(Callable)} tasks at a lower
+ * priority and with more limited slots.
  */
-public class ThreadPool {
+public class ThreadPool extends ThreadPoolExecutor {
 
     private final static Logger log = LoggerFactory.getLogger(ThreadPool.class);
 
-    private final LinkedBlockingQueue<Runnable> queue;
+    /**
+     * {@link Semaphore} protecting the <i>submission</i> of background tasks.
+     * The slot will not be made available though until <i>completion</i>, i.e.
+     * during {@link #afterExecute(Runnable, Throwable)}.
+     */
+    private final Semaphore maxBackground;
 
-    private final ThreadFactory factory;
+    /**
+     * Milliseconds to wait until calls to {@link #background(Callable)} will
+     * timeout.
+     */
+    private final long backgroundTimeout;
 
-    private final ExecutorService executor;
+    public ThreadPool() {
+        // Values from Executres.newCachedThreadPool
+        super(0, Integer.MAX_VALUE, 60L, TimeUnit.SECONDS,
+            new SynchronousQueue<Runnable>());
+        this.maxBackground = new Semaphore(10);
+        this.backgroundTimeout = 3600*1000;
 
-    public ThreadPool(int minThreads, int maxThreads, long msTimeout) {
-        queue = new LinkedBlockingQueue<Runnable>();
-        factory = null;
-        executor = new ThreadPoolExecutor(minThreads, maxThreads, msTimeout,
-                TimeUnit.MILLISECONDS, queue); // factory
     }
 
+    public ThreadPool(int minThreads, int maxThreads, long msTimeout,
+            int backgroundThreads, long backgroundTimeout) {
+        super(minThreads, maxThreads, msTimeout, TimeUnit.MILLISECONDS,
+                new LinkedBlockingQueue<Runnable>());
+        this.maxBackground = new Semaphore(backgroundThreads);
+        this.backgroundTimeout = backgroundTimeout;
+        log.info("ThreadPool: normal={}, background={}, timeout={}",
```

Comment:
may be confused with `msTimeout`

Reply:
I _knew_ I should have taken the time to clarify that... ;)